File: dotnet/src/Kajay.Core/KajayPatternCompiler.cs
using System.Diagnostics;

namespace Kajay;

internal sealed class KajayPatternCompiler
{
    private const int MaxCompiledStates = 4_096;
    private int _stateCount;

    public KajayPatternState? Compile(KajayPatternNode node)
    {
        try
        {
            return Compile(node, Create(new KajayAcceptState()));
        }
        catch (PatternLimitException)
        {
            return null;
        }
    }

    private KajayPatternState Compile(
        KajayPatternNode node,
        KajayPatternState next)
    {
        return node switch
        {
            KajayPatternNode.Empty => next,
            KajayPatternNode.Scalar scalar =>
                Create(new KajayMatchState(scalar.Pattern, next)),
            KajayPatternNode.Anchor anchor =>
                Create(new KajayAnchorState(anchor.IsStart, next)),
            KajayPatternNode.Sequence sequence => CompileSequence(sequence, next),
            KajayPatternNode.Alternation alternation => CompileAlternation(alternation, next),
            KajayPatternNode.Repeat repeat => CompileRepeat(repeat, next),
            _ => throw new UnreachableException(),
        };
    }

    private KajayPatternState CompileSequence(
        KajayPatternNode.Sequence sequence,
        KajayPatternState next)
    {
        KajayPatternState start = next;
        for (int index = sequence.Items.Count - 1; index >= 0; index -= 1)
        {
            start = Compile(sequence.Items[index], start);
        }

        return start;
    }

    private KajayPatternState CompileAlternation(
        KajayPatternNode.Alternation alternation,
        KajayPatternState next)
    {
        int last = alternation.Alternatives.Count - 1;
        KajayPatternState start = Compile(alternation.Alternatives[last], next);
        for (int index = last - 1; index >= 0; index -= 1)
        {
            start = Create(new KajayBranchState(
                Compile(alternation.Alternatives[index], next),
                start));
        }

        return start;
    }

    private KajayPatternState CompileRepeat(
        KajayPatternNode.Repeat repeat,
        KajayPatternState next)
    {
        KajayPatternState start = next;
        if (repeat.Maximum is null)
        {
            var branch = Create(new KajayBranchState(null, next));
            branch.First = Compile(repeat.Item, branch);
            start = branch;
        }
        else
        {
            for (int count = repeat.Maximum.Value; count > repeat.Minimum; count -= 1)
            {
                KajayPatternState optional = Compile(repeat.Item, start);
                start = Create(new KajayBranchState(optional, start));
            }
        }

        for (int count = 0; count < repeat.Minimum; count += 1)
        {
            start = Compile(repeat.Item, start);
        }

        return start;
    }

    private T Create<T>(T state)
        where T : KajayPatternState
    {
        _stateCount += 1;
        if (_stateCount > MaxCompiledStates)
        {
            throw new PatternLimitException();
        }

        return state;
    }

    private sealed class PatternLimitException : Exception;
}

File: packages/core/src/patterns/KajayPatternCompiler.ts
import type { PatternNode } from './KajayPatternNode.js';
import type { BranchState, PatternState } from './KajayPatternState.js';

const MAX_COMPILED_STATES = 4_096;

/** Compiles parsed patterns into a bounded Thompson state graph. */
export class KajayPatternCompiler {
  #stateCount = 0;

  compile(node: PatternNode): PatternState | undefined {
    try {
      const accept = this.#state<PatternState>({ kind: 'accept' });
      return this.#compile(node, accept);
    } catch (cause) {
      if (cause === PATTERN_LIMIT) {
        return undefined;
      }
      throw cause;
    }
  }

  #compile(node: PatternNode, next: PatternState): PatternState {
    switch (node.kind) {
      case 'empty':
        return next;
      case 'scalar':
        return this.#state({ kind: 'match', scalar: node.scalar, next });
      case 'anchor':
        return this.#state({ kind: 'anchor', edge: node.edge, next });
      case 'sequence':
        return node.items.reduceRight((tail, item) => this.#compile(item, tail), next);
      case 'alternation':
        return node.alternatives
          .map((alternative) => this.#compile(alternative, next))
          .reduceRight((second, first) => this.#state({ kind: 'branch', first, second }));
      case 'repeat':
        return this.#repeat(node.item, node.minimum, node.maximum, next);
    }
  }

  #repeat(
    item: PatternNode,
    minimum: number,
    maximum: number | undefined,
    next: PatternState,
  ): PatternState {
    let start = next;
    if (maximum === undefined) {
      const branch = this.#state<BranchState>({ kind: 'branch', second: next });
      branch.first = this.#compile(item, branch);
      start = branch;
    } else {
      for (let count = maximum; count > minimum; count -= 1) {
        const optional = this.#compile(item, start);
        start = this.#state({ kind: 'branch', first: optional, second: start });
      }
    }
    for (let count = 0; count < minimum; count += 1) {
      start = this.#compile(item, start);
    }
    return start;
  }

  #state<T extends PatternState>(state: T): T {
    this.#stateCount += 1;
    if (this.#stateCount > MAX_COMPILED_STATES) {
      throw PATTERN_LIMIT;
    }
    return state;
  }
}

const PATTERN_LIMIT = Symbol('Kajay pattern compiled-state limit');
